feature/desul-atomics (#1073)
LLNL/RAJA: docs/sphinx/user_guide/feature/atomic.rst

    * CUDA native 64-bit double `atomicAdd` is used.

---------------------
DESUL Atomics Support
---------------------

> RAJA supports the use of `DESUL Atomics <https://github.com/desul/desul>`_ as

Reviewer: ```suggestion
RAJA provides support for the use of `DESUL Atomics <https://github.com/desul/desul>`_ as
```

> an alternative backend to the default implementation of RAJA atomics. Enabling

Reviewer: ```suggestion
an alternative backend to the default implementation of RAJA atomics. DESUL atomics are considered an **experimental** feature in RAJA at this point. DESUL atomics
```

> this feature may impact the performance of some atomic functions. While switching

Reviewer: ```suggestion
may impact the performance of some atomic functions. While switching
```

> to DESUL typically yields positive or neutral performance results, some atomic

Reviewer: ```suggestion
to DESUL atomics typically yields positive or neutral performance results, some atomic
```

operations may perform worse when using DESUL.

To enable DESUL Atomics:

#. Ensure that RAJA and its dependencies are configured to use C++14.
#. Set ``RAJA_ENABLE_DESUL_ATOMICS=On``.

Enabling DESUL Atomics alters RAJA atomic functions to be wrapper-functions for their

> DESUL counterparts. This removes any code changes necessary for switching between

Reviewer: ```suggestion
DESUL counterparts. This removes the need for user code changes to switch between
```

> DESUL and RAJA implementations, except for cases where RAJA atomic helper functions

Reviewer: ```suggestion
DESUL and RAJA implementations. The exception to this is when RAJA atomic helper functions
```

> are used instead of the backwards-compatible API functions specified by :ref:`atomic-ops`.

Reviewer: ```suggestion
are used instead of the backwards-compatible API functions specified by :ref:`atomic-ops`. By *helper functions*, we mean the RAJA atomic methods which take a reduction policy object as the first argument, instead of specifying the reduction policy type as a template parameter. 
```

> DESUL compiles atomic functions in accordance to the scope in which they are

Reviewer: ```suggestion
DESUL atomic functions are compiled with the proper back-end implementation based on the scope in which they are
```

> called, this removes the requirement of using atomic policies to specify

Reviewer: ```suggestion
called, which removes the need to specify atomic policies for
```

> target backends. As a result, atomic policies such as ``cuda_atomic`` or ``omp_atomic``

Reviewer: ```suggestion
target back-ends. As a result, atomic policies such as ``cuda_atomic`` or ``omp_atomic``
```

> will be ignored when DESUL is enabled, but are still necessary to pass in as parameters

Reviewer: ```suggestion
are ignored when DESUL is enabled, but are still necessary to pass in as parameters
```

> to the RAJA API.

Reviewer: ```suggestion
to the RAJA API. This will likely change in the future and RAJA atomic policies will be removed.
```
